### PR TITLE
WIP: Update MAPL and other GEOS-ESM packages

### PR DESCRIPTION
**Pull request by mathomp4**
This is an update to the MAPL and other GEOS-ESM packages in NOAA-EMC/spack. 

The biggest changes would be in MAPL. I removed the `ESMA_GFE_NAMESPACE` option as it is nigh-ancient and I'm not sure any version here needs it (might be wrong, though.) I also added dependencies for yafyaml when building with ExtData2G support. There was an interface change at MAPL 2.23 so you need a newer yafyaml. (This also triggers a split in pflogger.)

cc: @climbfuji @tclune

---

Changed:

* mapl
* esmf (slight bug fix)
* FMS (need a new variant)

Additions:

* flap (A weird one. flap is also in canonical spack, but we use a fork of it. This can probably never go back to mainline spack. I've hesitated pushing our changes to the upstream flap repo as MAPL eventually will move to fargparse or, most likely, move away from command line arguments altogether and go to YAML.)
* geosgcm (testing at the moment)
* mepo (for geosgcm, though maybe it should go away in favor of tarfiles?)

---

Testing:

* Building:
  - [x] Does it build in spack-stack?
* UFS
  - [ ] Do the libraries it builds, build with MAPL-in-UFS?
  - [ ] Does the MAPL in this PR work with UFS?
  - [ ] Can it run?
* GEOSgcm
  * Internal MAPL and Internal FMS
    - [ ] Does it build GEOSgcm?
    - [ ] Does GEOSgcm run?
  * Internal MAPL and Spack FMS
    - [ ] Does it build GEOSgcm?
    - [ ] Does GEOSgcm run?
  * Spack MAPL and Internal FMS
    - [ ] Does it build GEOSgcm?
    - [ ] Does GEOSgcm run?
  * Spack MAPL and Spack FMS
    - [ ] Does it build GEOSgcm?
    - [ ] Does GEOSgcm run?

### CODE REVIEW

**Comment by mathomp4** [comment]
A few updates to remind myself. I have gotten, I think, a good set of libraries. I did have to do a few things.

In the `common/packages.yaml` file for my empty environment I added:
```yaml
     pflogger:
       variants: +mpi
     pfunit:
       variants: +mpi+fhamcrest
```
to lock in some variants we want.

I also *really* only want ESMF 8.3.0 but I just could not get spack to build it. The only way was to comment out the 8.3.0b09 line here:

https://github.com/NOAA-EMC/spack/blob/1248dfd679e9ee8095222b34b0a71217c52c272c/var/spack/repos/builtin/packages/esmf/package.py#L24-L34

because spack seemed to think 8.3.0b09 was *newer* than 8.3.0. Maybe @climbfuji knows a better way, but I'm at the "throw things at the wall" stage 😄 

But I was able to build:
```
spack install mapl@2.26.0+extdata2g+flap+pflogger
```

I'm currently trying to build GEOSgcm...but not with spack MAPL, but instead an internal MAPL. I'm also going to try both a internal-FMS as well as an FMS-from-spack. But to do that, I'll need to slightly fix up the FMS here in this repo, I think. I'll report back!

**Comment by tclune** [comment]
> A few updates to remind myself. I have gotten, I think, a good set of libraries. I did have to do a few things.
> 
> In the `common/packages.yaml` file for my empty environment I added:
> 
> ```yaml
>      pflogger:
>        variants: +mpi
>      pfunit:
>        variants: +mpi+fhamcrest
> ```

I wonder if pfunit should have an openmp variant as well?

**Comment by mathomp4** [comment]
> > A few updates to remind myself. I have gotten, I think, a good set of libraries. I did have to do a few things.
> > In the `common/packages.yaml` file for my empty environment I added:
> > ```yaml
> >      pflogger:
> >        variants: +mpi
> >      pfunit:
> >        variants: +mpi+fhamcrest
> > ```
> 
> I wonder if pfunit should have an openmp variant as well?

It does have one:
```
    variant("openmp", default=False, description="Enable OpenMP")
```
I just didn't turn it on by default. We might want to? I just figure I usually build it with `--DSKIP_OPENMP=ON` and...well crap. Now I think the package doesn't handle that correctly. Let me take a look...

ETA: Ah it does:

```python
        if spec.satisfies("@4.0.0:"):
            args.append("-DSKIP_MPI=%s" % ("YES" if "~mpi" in spec else "NO"))
            args.append("-DSKIP_OPENMP=%s" % ("YES" if "~openmp" in spec else "NO"))
            args.append("-DSKIP_FHAMCREST=%s" % ("YES" if "~fhamcrest" in spec else "NO"))
            args.append("-DSKIP_ESMF=%s" % ("YES" if "~esmf" in spec else "NO"))
        else:
            args.append(self.define_from_variant("MPI", "mpi"))
            args.append(self.define_from_variant("OPENMP", "openmp"))
```
I was just looking in the wrong place.

**Comment by tclune** [comment]
@mathomp4 OK - glad you are on it.   Agree that we want it off for now; I was just keying off the absence.   So ignore what I wrote.

**Comment by mathomp4** [comment]
Okay. Spack has accepted the GFE packages in:

https://github.com/spack/spack/pull/34476

**Comment by mathomp4** [comment]
For my own "records" (as it were), I've added some changes from @bd4 that were proposed to spack mainline here: https://github.com/spack/spack/pull/34713

I'm not sure when that PR will get in, but I want to make sure I capture the changes here.

**Comment by climbfuji** [comment]
@mathomp4 FYI the spack-stack and spack repos will be moving to JCSDA after the 1.3.0 release was created (in approx 2 weeks?). While I don't foresee any problems with open PRs, it might be easier if we try to merge currently open PRs before then. What is the status of this PR?

**Comment by mathomp4** [comment]
> @mathomp4 FYI the spack-stack and spack repos will be moving to JCSDA after the 1.3.0 release was created (in approx 2 weeks?). While I don't foresee any problems with open PRs, it might be easier if we try to merge currently open PRs before then. What is the status of this PR?

@climbfuji Probably stupid out of date. I won't have time to look at it today, but I'll try next week. My guess is I'll probably just delete this branch and start anew (since all of GFE is in mainline spack now). Shouldn't be too hard as the MAPL package is really the big one now. I'll probably focus on it and then eventually bring in GEOSgcm (and maybe mepo?)

**Comment by mathomp4** [comment]
@climbfuji So, I updated the files here and I was able to install MAPL. So, yay. I then wanted to test building geosgcm but I saw I needed a newer ESMF. Fine:
```
$ spack install esmf@8.4.0
==> Error: Package 'mapl' not found.
You may need to run 'spack clean -m'.
```
Huh. I'm new at all this stack/environment stuff. Any idea what I might have done wrong for spack to say this?

ETA: Oh. wait:
```
$ spack find esmf
==> In environment /gpfsm/dnb44/mathomp4/spack-stack/envs/geosgcm.discover
==> Root specs
geosgcm  mapl@spacktest  mapl@2.31.0 +extdata2g+fargparse+flap+pflogger+pfunit  mapl@2.34.3  mapl@develop  mepo  py-pyyaml

==> Installed packages
-- linux-sles12-skylake_avx512 / intel@2022.0.1 -----------------
esmf@8.4.0
==> 1 installed package
```
Already installed. Maybe I need to do some weird concretize things?

**Comment by climbfuji** [comment]
spack clean -a doesn’t help?

**Comment by mathomp4** [comment]
> spack clean -a doesn’t help?

Nope. But I'm going to try a few things. Worse comes to worst, I just reclone!

**Comment by climbfuji** [comment]
> > spack clean -a doesn’t help?
> 
> Nope. But I'm going to try a few things. Worse comes to worst, I just reclone!

If the spack version is too different, you may have to re-bootstrap from a fresh clone

**Comment by mathomp4** [comment]
Well, I think my environment was just *not* good. I did a:
```
spack concretize -Uf
```
as that often fixes things and things changed. A lot. I think I just need to start afresh.

**Comment by mathomp4** [comment]
@climbfuji I am still working on this and found a bug in my (every?) MAPL version if you try and build `+pfunit`. I think I have a fix, but the fun part is testing the fix. I added a new branch for MAPL, and the weird thing is, spack keeps wanting to patch it. But in my MAPL `package.py` I've commented out *every* patch command I see...and it's still trying to patch!

```
==> Installing mapl-spack-eprnncn4hdrgc6bllawpudthctbt7qhe
==> mapl exists in binary cache but with different hash
==> No binary for mapl-spack-eprnncn4hdrgc6bllawpudthctbt7qhe found: installing from source
==> Error: NoSuchPatchError: Couldn't find patch for package builtin.mapl with sha256: 7b9c28b9343403b0d1d8e58478982e575b839a811db3c95f0b3987dcfd841287

/gpfsm/dnb44/mathomp4/spack-stack/spack/lib/spack/spack/package_base.py:1456, in do_fetch:
       1453            self.stage.check()
       1454
       1455        self.stage.cache_local()
  >>   1456
       1457        for patch in self.spec.patches:
       1458            patch.fetch()
       1459            if patch.stage:
```

Any idea where this might be coming from?

**Comment by climbfuji** [comment]
You need to re-concretize for spack to see changes regarding patches

**Comment by mathomp4** [comment]
> You need to re-concretize for spack to see changes regarding patches

Ahhh. Okay. I keep forgetting about that! On to other fun not using a version number!

**Comment by mathomp4** [comment]
@AlexanderRichert-NOAA I think I finally have the MAPL package here in good shape. If you have a chance, can you try it out with your code? If it works, I can undraft this PR and even add it to spack mainline for you.

**Comment by AlexanderRichert-NOAA** [comment]
Awesome, I'll test it out ASAP.

**Comment by mathomp4** [comment]
A note to @climbfuji. The `pfunit` here will of course be superseded by the newer better one in current `develop` spack. Same for `fms`. I've only added FMS here because this doesn't have the new "precision" one.

I'm debating if `mepo`, `geosgcm` and `flap` need to be here or not. I'm keeping them here just in case. Eventually FLAP use in GEOS is going away, and I'm pretty sure my GEOSgcm package doesn't need mepo anymore.

File: var/spack/repos/builtin/packages/fms/package.py [hunk 1]
```diff
@@ -59,6 +59,11 @@ class Fms(CMakePackage):
         git = "https://github.com/JCSDA/fms.git"
     # *DH 20220602
 
+    variant(
```
Comment (climbfuji) [inline var/spack/repos/builtin/packages/fms/package.py:62]: Instead of this we should pull in  the updates from https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/fms/package.py (can be a separate PR so you don't have to deal with it)

File: var/spack/repos/jcsda-emc/packages/geosgcm/package.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    depends_on("pflogger@1.9.1:")
+    depends_on("fargparse@1.4.1:")
+    depends_on("flap@geos")
+
+    # MAPL as library would be like:
+    #  depends_on("mapl@2.34:+flap+pflogger+extdata2g+fargparse")
+    # but we don't want to do this in general due to the speed of MAPL development
+
+    # When we move to FMS as library, we'll need to add this:
+    #depends_on("fms@2022.04:~gfs_phys+fpic~quad_precision+32bit+64bit+yaml constants=GEOS")
```
Comment (climbfuji) [inline var/spack/repos/jcsda-emc/packages/geosgcm/package.py:72]: When the time comes, we'll need a way to describe the different variants in the module names (not now!), e.g. `fms/2022.04-geos` based on the constants value).

File: var/spack/repos/builtin/packages/fms/package.py (hunk 1 above)
Comment (climbfuji) [reply var/spack/repos/builtin/packages/fms/package.py:62]: I can do this if you like

**Review by climbfuji — CHANGES_REQUESTED** [review]

**Comment by AlexanderRichert-NOAA** [comment]
This worked for me on hera with UFS (ea0b6e4) using ufs-weather-model-static. The only change needed was setting `fms +32bit`.

File: var/spack/repos/builtin/packages/fms/package.py (hunk 1 above)
Comment (mathomp4) [reply var/spack/repos/builtin/packages/fms/package.py:62]: @climbfuji If you can, please. I have a [PR in to spack mainline](https://github.com/spack/spack/pull/35968) to update some other GFE packages. 

It's also possible we might want to wait for the [ESMF PR from @theurich](https://github.com/spack/spack/pull/35980) before pulling in spack? Not sure.

**Comment by mathomp4** [comment]
> This worked for me on hera with UFS (ea0b6e4) using ufs-weather-model-static. The only change needed was setting `fms +32bit`.

Ooh. Nice!

File: var/spack/repos/builtin/packages/fms/package.py (hunk 1 above)
Comment (mathomp4) [reply var/spack/repos/builtin/packages/fms/package.py:62]: Also, I will say I tried using the newer spack `maintainers()` function in this PR but spack wasn't happy. It's possible my spack was old?

Comment (AlexanderRichert-NOAA) [reply var/spack/repos/builtin/packages/fms/package.py:62]: We haven't yet pulled that functionality from main spack, so with NOAA-EMC spack it will yell at you about the function being undefined :)

**Comment by mathomp4** [comment]
Closing. Will make a new PR with geosgcm (and maybe mepo and flap) when #241 is pulled in and tested.
